### ISSUE TITLE
Add a phase that removes ResolvedTypeRefs before codegen

### ISSUE THREAD

**Issue by fitzgen** [servo/rust-bindgen#534]
Ever since we got the graphviz visualizations of our IR graph, it has become immediately clear that we have a TON of `ResolvedTypeRef`s all over the place. In fact, 8 out of 15 IR items (more than half!) are `ResolvedTypeRef`s in this example I happen to be looking at:

![using](https://cloud.githubusercontent.com/assets/74571/23174809/6c31bef6-f812-11e6-808d-83cd3bb694a4.png)

A pass over the IR items where we replace all references to `ResolvedTypeRef`s with references to the type being transitively referenced via the `ResolvedTypeRef` would probably give us speed ups in codegen and analysis, because we wouldn't have to look at useless items anymore. It would also make looking at these graphs easier, since there would be less visual noise :-P

A good time to do this would be in `ir::context::BindgenContext::gen`, after we resolve unresolved type refs in `resolve_typerefs`. https://github.com/servo/rust-bindgen/blob/master/src/ir/context.rs#L529

I can mentor whoever would like to take a stab at this.

**Comment by fitzgen** [comment]
If you want to pick this up, leave a comment and I can assign you :)

**Comment by samliu** [comment]
Hey Nick! (How are you?!) I've been learning Rust and this looks really interesting. I'd be down to take a stab :)

**Comment by fitzgen** [comment]
Great! Let me know if you have any questions or anything like that.

I just realized that fully removing `ResolvedTypeRef`s would require modifying `ir::traversal::Trace` to be generic over either `&mut ItemId` edges or just `ItemId`s like it is now. This is a bit more tricky than I anticipated, and so I think we should resign ourselves to flattening `ResolvedTypeRef` chains to a single `ResolvedTypeRef`. That is,

```
ResolvedTypeRef -> ResolvedTypeRef -> ResolvedTypeRef -> SomeType
```

should become:

```
ResolvedTypeRef -> SomeType
```

The core of this new phase should look something like this pseudo code:

```rust
impl BindgenContext {
    // ...

    fn flatten_resolved_type_ref(&mut self) {
        for (id, item) in self.items {
            if item is a resolved type ref {
                self.flatten_one_resolved_type_ref(id);
            }
        }
    }

    // Recursively walk the chain of ResolvedTypeRefs and flatten it as we
    // come back out of the recursion.
    fn flatten_one_resolved_type_ref(&mut self, id: ItemId) -> ItemId { 
        if id's item is a resolved type ref {
            let final_referenced_id = self.flatten_one_resolved_type_ref(item's referenced type);
            item's referenced type = final_referenced_id;
            final_referenced_id
        } else {
            id
        }
    }

    // ...
}
```

Bonus points for keeping a debug-only set of items we've walked in the recursive function and asserting that we don't get a cycle :) See https://github.com/servo/rust-bindgen/blob/master/src/ir/item.rs#L69-L87

**Comment by fitzgen** [comment]
(Also, woah! Hey! Didn't read your username very carefully, at first! :-P)

**Comment by emilio** [comment]
I have an issue with this, and is that this would lose the constness information, which we do use for code generation.

**Comment by fitzgen** [comment]
> I have an issue with this, and is that this would lose the constness information, which we do use for code generation.

Crap, I forgot about that. It would be great if we separated the concept of a resolved type reference from a `const <type>`, then we could move forward here.

@samliu interested in this prerequisite groundwork?

**Comment by samliu** [comment]
Definitely!

**Comment by emilio** [comment]
As an idea, we probably want to call this something like `QualifiedType`,
and it'd be something like:

```rust
struct Qualifiers {
    is_const: bool,
}

struct QualifiedType {
    ty: ItemId,
    qualifiers: Qualifiers,
}
```

**Comment by fitzgen** [comment]
Yep, exactly what I was imagining!

It will also need to be `Trace`d from the `impl Trace for Type` as well (either directly there, or as a second `Trace` implementation that is called from there).

**Comment by fitzgen** [comment]
(And there would be a new `TypeKind` variant called `TypeKind::Qualified(QualifiedType)` or something)

**Comment by emilio** [comment]
> (And there would be a new TypeKind variant called TypeKind::Qualified(QualifiedType) or something)

Why would that be needed? We only care about qualifiers in top-level types, right?

I.e., `const StructName* foo;` would be a `TypeKind::Pointer(QualifiedType)`, and so on with fields, arrays, etc.

Am I missing something?

**Comment by emilio** [comment]
> Why would that be needed? We only care about qualifiers in top-level types, right?

With top-level types, I mean in the stuff that holds from `TypeKind`, `Var`, etc.

**Comment by fitzgen** [comment]
I don't understand the distinction you are trying to make, sorry if I'm being thick...

**Comment by fitzgen** [comment]
> I.e., const StructName* foo; would be a TypeKind::Pointer(QualifiedType), and so on with fields, arrays, etc.

Aha, I see where our imaginations are diverging.

I would imagine `const Structname*` to result in IR like

```
ItemId(0) = Item(TypeKind::Pointer(ItemId(1)))
ItemId(1) = Item(TypeKind::Qualified(ItemId(2)))
ItemId(2) = Item(TypeKind::Comp(StructName))
```

**Comment by emilio** [comment]
Right, I was imagining making `QualifiedType` a thin wrapper on top of an `ItemId`. I think it's worth for a few reasons:

 * First, you don't need all the memory impact of an `Item` to just indicate that a type is `const`.
 * Second, I think the qualifications are orthogonal to the type, so I think having a type being a qualified type is confusing. Also, we'd need to guarantee that a `TypeKind::Qualified` wouldn't point to another of the same kind, etc.
 * Third, the approach you're suggesting makes us having to create two items in common paths that now create one (like pointer construction), which would be a bigger refactor than needed IMO.
 * Finally, clang seems to agree with me in this regard, and it uses internally a structure similar to this (see [the `QualType` definition](https://github.com/llvm-mirror/clang/blob/f5802a7c28a2be35565dc66a16679ba383f4dfd2/include/clang/AST/Type.h#L599)).

WDYT? I think representing a struct field like:

```c
struct foo {
    const struct bar* const bar_field;
};
```

As:

```
QualifiedType(ItemId(1), const)
```

where:

```
ItemId(1): TypeKind::Pointer(QualifiedType(ItemId(2), const))
ItemId(2): TypeKind::Comp(..)
```

Is more appealing than using the already arguably somewhat bloated `Type`.

**Comment by fitzgen** [comment]
Ok, yeah that sounds well-motivated.

**Comment by fitzgen** [comment]
@samliu any progress? any questions I can help answer? :)

**Comment by samliu** [comment]
Sorry no progress :( haven't had time to get to this, out of town past two
weekends.

What is a good timeline for addressing issues? I definitely want to do this
but probably need to prioritize better.

**Comment by fitzgen** [comment]
No worries!

There's no rush with this, so as long as you still plan on tackling it, we can keep it assigned. If for whatever reason you don't want to or can't work on this anymore, then we can unassign and let someone else grab it.

**Comment by samliu** [comment]
Thanks Nick. I'll do my best to make progress in a timely fashion
nonetheless, or give it up if I can't commit to that anymore.

**Comment by RReverser** [comment]
Was just about to report this, but looks like a known issue. Is there anything we could help with?

**Comment by fitzgen** [comment]
@RReverser if you wanted to pick this up, I think that's fine, given that there's been silence for a while.

@samliu you're not still planning on doing this, are you?

**Comment by samliu** [comment]
Sorry yeah I won't be able to get to this :( feel free to reassign
